File: src/fable-library/RegExp.ts
export type MatchEvaluator = (match: any) => string;

export function create(pattern: string, options: number = 0) {
  // Supported RegexOptions
  // * IgnoreCase:  0x0001
  // * Multiline:   0x0002
  // * Singleline:  0x0010
  // * ECMAScript:  0x0100 (ignored)
  if ((options & ~(1 ^ 2 ^ 16 ^ 256)) !== 0) {
    throw new Error("RegexOptions only supports: IgnoreCase, Multiline, Singleline and ECMAScript");
  }
  let flags = "g";
  flags += options & 1 ? "i" : ""; // 0x0001 RegexOptions.IgnoreCase
  flags += options & 2 ? "m" : "";
  flags += options & 16 ? "s" : "";
  return new RegExp(pattern, flags);
}
// From http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex

export function escape(str: string) {
  return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
}

export function unescape(str: string) {
  return str.replace(/\\([\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|])/g, "$1");
}

export function isMatch(str: string | RegExp, pattern: string, options: number = 0) {
  let reg: RegExp;
  reg = str instanceof RegExp
    ? (reg = str as RegExp, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  return reg.test(str as string);
}

export function match(str: string | RegExp, pattern: string, options: number = 0) {
  let reg: RegExp;
  reg = str instanceof RegExp
    ? (reg = str as RegExp, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  return reg.exec(str as string);
}

export function matches(str: string | RegExp, pattern: string, options: number = 0) {
  let reg: RegExp;
  reg = str instanceof RegExp
    ? (reg = str as RegExp, str = pattern, reg.lastIndex = options, reg)
    : reg = create(pattern, options);
  if (!reg.global) {
    throw new Error("Non-global RegExp"); // Prevent infinite loop
  }
  let m = reg.exec(str as string);
  const matches: RegExpExecArray[] = [];
  while (m !== null) {
    matches.push(m);
    m = reg.exec(str as string);
  }
  return matches;
}

export function options(reg: RegExp) {
  let options = 256; // ECMAScript
  options |= reg.ignoreCase ? 1 : 0;
  options |= reg.multiline ? 2 : 0;
  return options;
}

export function replace(
  reg: string | RegExp, input: string,
  replacement: string | MatchEvaluator,
  limit?: number, offset: number = 0): string {
  function replacer() {
    let res = arguments[0];
    if (limit) {
      limit--;
      const match: any = [];
      const len = arguments.length;
      for (let i = 0; i < len - 2; i++) {
        match.push(arguments[i]);
      }
      match.index = arguments[len - 2];
      match.input = arguments[len - 1];
      res = (replacement as MatchEvaluator)(match);
    }
    return res;
  }
  if (typeof reg === "string") {
    const tmp = reg as string;
    reg = create(input, limit ?? 0);
    input = tmp;
    limit = undefined;
  }
  if (typeof replacement === "function") {
    limit = limit == null ? -1 : limit;
    return input.substring(0, offset) + input.substring(offset).replace(reg as RegExp, replacer);
  } else {
    // $0 doesn't work with JS regex, see #1155
    replacement = replacement.replace(/\$0/g, (_s) => "$&");
    if (limit != null) {
      let m: RegExpExecArray;
      const sub1 = input.substring(offset);
      const _matches = matches(reg, sub1);
      const sub2 = matches.length > limit ? (m = _matches[limit - 1], sub1.substring(0, m.index + m[0].length)) : sub1;
      return input.substring(0, offset) + sub2.replace(reg as RegExp, replacement as string)
        + input.substring(offset + sub2.length);
    } else {
      return input.replace(reg as RegExp, replacement as string);
    }
  }
}

export function split(reg: string | RegExp, input: string, limit?: number, offset: number = 0) {
  if (typeof reg === "string") {
    const tmp = reg as string;
    reg = create(input, limit ?? 0);
    input = tmp;
    limit = undefined;
  }
  input = input.substring(offset);
  return input.split(reg as RegExp, limit);
}

File: src/fable-library-dart/RegExp.dart
// ignore_for_file: file_names
import "./String.dart" as string_mod;

class FailMatch implements Match {
  @override
  String? operator [](int group) => null;

  @override
  int get end => 0;

  @override
  String? group(int group) => null;

  @override
  int get groupCount => 0;

  @override
  List<String?> groups(List<int> groupIndices) => const [];

  @override
  String get input => '';

  @override
  Pattern get pattern => '';

  @override
  int get start => 0;
}

class GroupIterable extends Iterable<String?> {
  final Match match;
  GroupIterable(this.match);

  @override
  get iterator => GroupIterator(match);
}

class GroupIterator implements Iterator<String?> {
  final Match match;
  int _nextIndex = 0;

  GroupIterator(this.match);

  @override
  String? current;

  @override
  bool moveNext() {
    if (_nextIndex > match.groupCount) {
      return false;
    }
    current = match.group(_nextIndex++);
    return true;
  }
}

RegExp create(String pattern, [int options = 0]) {
  // Supported RegexOptions
  // * IgnoreCase:  0x0001
  // * Multiline:   0x0002
  // * Compiled:    0x0008 (ignored)
  // * Singleline:  0x0010
  // * ECMAScript:  0x0100 (ignored)
  if ((options & ~(1 ^ 2 ^ 8 ^ 16 ^ 256)) != 0) {
    throw Exception(
        'RegexOptions only supports: IgnoreCase, Multiline, Compiled, Singleline and ECMAScript');
  }
  return RegExp(pattern,
      unicode: true,
      multiLine: options & 2 > 0,
      caseSensitive: options & 1 == 0,
      dotAll: options & 16 > 0);
}

// From http://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex

String escape(String str) {
  return RegExp.escape(str);
  // return str.replaceAllMapped(
  //   RegExp(r'[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]'),
  //   (m) => '\\${m.group(0)}',
  // );
}

String unescape(String str) {
  return str.replaceAllMapped(
    RegExp(r'\\([\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|])'),
    (m) => m.group(1)!,
  );
}

bool isMatch(RegExp reg, String input, [int startAt = 0]) {
  return startAt > 0
      ? reg.allMatches(input, startAt).isNotEmpty
      : reg.hasMatch(input);
}

String matchValue(Match value) {
  return value.group(0) ?? '';
}

String groupValue(String? group) {
  return group ?? '';
}

int matchLength(Match match) {
  return match.end - match.start;
}

int groupLength(String? group) {
  return group?.length ?? 0;
}

bool matchSuccess(Match match) {
  return match is! FailMatch;
}

bool groupSuccess(String? group) {
  return group != null;
}

String? matchGroup(Match match, int index) {
  return match.group(index);
}

String? matchNamedGroup(Match match, String index) {
  if (match is RegExpMatch) {
    try {
      return match.namedGroup(index);
    } catch (_) {
      return null;
    }
  } else {
    return null;
  }
}

Match match(RegExp reg, String input, [int? startAt]) {
  if (startAt != null && startAt > 0) {
    // Note that reg.matchAsPrefix behaves as sticky regex in JS
    final iterator = reg.allMatches(input, startAt).iterator;
    return iterator.moveNext() ? iterator.current : FailMatch();
  } else {
    return reg.firstMatch(input) ?? FailMatch();
  }
}

int options(RegExp reg) {
  var options = 256; // ECMAScript
  options |= !reg.isCaseSensitive ? 1 : 0;
  options |= reg.isMultiLine ? 2 : 0;
  options |= reg.isDotAll ? 16 : 0;
  return options;
}

String _replaceWith(
    Pattern pattern, String input, String Function(Match) replace,
    [int? limit]) {
  var count = 0;
  pattern = pattern is String ? create(pattern) : pattern;
  return input.replaceAllMapped(pattern, (match) {
    count++;
    if ((limit != null && count > limit)) {
      return match.group(0)!;
    } else {
      return replace(match);
    }
  });
}

String replaceWith(
        Pattern pattern, String input, String Function(Match) replace,
        [int? limit, int offset = 0]) =>
    offset > 0
        ? input.substring(0, offset) +
            _replaceWith(pattern, input.substring(offset), replace, limit)
        : _replaceWith(pattern, input, replace, limit);

final _replaceMacros = RegExp(r'(?<!\$)\$(\d+|\{\w+\})');

String replace(Pattern pattern, String input, String replace,
    [int? limit, int offset = 0]) {
  if (_replaceMacros.hasMatch(replace)) {
    return replaceWith(pattern, input, (match) {
      return replace.replaceAllMapped(_replaceMacros, (rep) {
        final index = rep.group(1)!;
        if (index.startsWith('{')) {
          final name = index.substring(1, index.length - 1);
          return matchNamedGroup(match, name) ?? rep.group(0)!;
        } else {
          return match.group(int.parse(index)) ?? rep.group(0)!;
        }
      });
    }, limit, offset);
  } else if (limit == null) {
    pattern = pattern is String ? create(pattern) : pattern;
    return offset > 0
        ? input.substring(0, offset) +
            input.substring(offset).replaceAll(pattern, replace)
        : input.replaceAll(pattern, replace);
  } else {
    return replaceWith(pattern, input, (_) => replace, limit, offset);
  }
}

List<String> splitWithPattern(String input, String pattern, [int? options]) {
  final reg = create(pattern, options ?? 0);
  return input.split(reg);
}

List<String> split(RegExp pattern, String input, [int? limit, int? offset]) {
  input = offset == null ? input : input.substring(offset);
  return limit == null
      ? input.split(pattern)
      : string_mod.splitWithRegExp(input, pattern, limit);
}
